tools/postgres/diff_schema.js

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env bash
+var program = require('commander');
+var randomstring = require("randomstring");
+const execSync = require('child_process').execSync;
+var path = require('path');
+var fs = require('fs');
+var glob = require("glob")
+var tmp = require('tmp');
+
+let POSTGRES_URL = process.env.POSTGRES_URL
+let ORIGINAL_POSTGRES_URL = (typeof POSTGRES_URL !== 'undefined') ? POSTGRES_URL:"jdbc:postgresql://localhost/webknossos"
+const scriptdir = __dirname
+const scriptName = __filename
+
+function dump(parameter) {
+	var cleanUp = function(){};
+	if(parameter == "DB") { 
+		POSTGRES_URL=ORIGINAL_POSTGRES_URL // this environment variable is passed to dump_schema.sh
+	} else {
+		const tempDbName = generateRandomName()
+		const postgresDirname = path.dirname(ORIGINAL_POSTGRES_URL)
+		POSTGRES_URL= postgresDirname + '/' + tempDbName;
+		const dbName = execSync(scriptdir+'/db_name.sh', {env: {'POSTGRES_URL': POSTGRES_URL}}).toString().trim() // "trim" to remove the line break
+		if(dbName !== tempDbName) {
+			console.log("Wrong dbName")	
+			process.exit()
+		}
+		console.log("Creating DB " + dbName)
+		const dbHost = execSync(scriptdir+'/db_host.sh', {env: {'POSTGRES_URL': POSTGRES_URL}}).toString().trim()
+		execSync(`psql -U postgres -h ${dbHost} -c "CREATE DATABASE ${dbName};"`, {env: {'PGPASSWORD': 'postgres'}})
+		var fileNames = glob.sync(parameter)
+		let concatenateFileNames = ""
+		fileNames.forEach(function(fileName) {
+  			concatenateFileNames += " -f " + fileName;
+		});
+		execSync(`psql -U postgres -h ${dbHost} --dbname="${dbName}" -v ON_ERROR_STOP=ON -q ${concatenateFileNames}`, {env: {'PGPASSWORD': 'postgres'}})
+		cleanUp = function(){
+			console.log(`CLEANUP: DROP DATABASE ${dbName}`)
+			execSync(`psql -U postgres -h ${dbHost} -c \"DROP DATABASE ${dbName};\"`, {env: {'PGPASSWORD': 'postgres'}})
+		}
+	}
+	var tmpDir = tmp.dirSync();
+	try {
+		execSync(`${scriptdir}/dump_schema.sh ${tmpDir.name}`, {env: {'POSTGRES_URL': POSTGRES_URL}})
+	} catch(err) {
+		console.log(`CLEANUP: remove ${tmpDir.name}`)
+		execSync(`rm -rf ${tmpDir.name}`)
+		process.exit(1)
+	}
+	cleanUp();
+	return tmpDir.name
+}
+
+function generateRandomName(){
+	random = randomstring.generate({
+  				length: 8,
+  				charset: 'alphanumeric',
+  				capitalization: 'lowercase'
+			})
+	return 'wk_tmp_' + random
+}
+
+process.on('exit', function(code) {
+    if (dir1) {
```

Comment:
Is it enough to check if the variable is set or do we need to be really careful and check if the directory itself exists?

Reply:
Checking if they are defined should be enough, as `rm -f` does not fail if the directory does not exist. There's still one issue: this statement does not check if they are defined, but if they are empty.

So you should either initialize them before with `false` or an empty String, or use something like `if (typeof dir1 !== 'undefined' && dir1)`

Reply:
I don't think you should rely on the `exit` event, but wrap your code in a big try/catch/finally